### PR TITLE
OPAL-549

### PR DESCRIPTION
**Pull request by JA-USAF**
Add pantry factory

### CODE REVIEW

File: weave/tests/test_pantry_factory.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+# Create Index CONSTRUCTORS of Indexes to be tested, and add to indexes list.
+indexes = [IndexPandas, IndexSQLite]
+indexes_ids = ["Pandas", "SQLite"]
+
+# Only add IndexSQL if the env variables are set and dependencies are present.
+if "WEAVE_SQL_HOST" in os.environ and "sqlalchemy" in sys.modules:
+    indexes.append(IndexSQL)
+    indexes_ids.append("SQL")
+
+# Create combinations of the above parameters to pass into the fixture..
```
Comment (AM-USAF) [inline weave/tests/test_pantry_factory.py:40]: Extra period?

**Review by hh-usaf — DISMISSED** [review]
I was able to make a pantry the old way and with a config file using create_pantry

File: weave/pantry_factory.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+            kwargs.pop("config_file"),
+            "r",
+            encoding="utf-8"
+        ) as config_file:
+            config = json.load(config_file)
+
+        pantry = _create_pantry_from_config(config, **kwargs)
+    # Create a pantry using a config that exists inside the pantry directory.
+    elif "pantry_path" in kwargs:
+        file_system = kwargs.pop("file_system")
```
Comment (AM-USAF) [inline weave/pantry_factory.py:65]: I get an error here if I only pass in an index and pantry path and no file system

File: weave/pantry_factory.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    config: dict
+        Dictionary that contains the contents of the config.json file.
+    Optional kwargs passed to the Index constructor.
+
+    Returns:
+    -----------
+    weave.Pantry object
+    """
+    # Parse the index type used by this pantry.
+    index_type = config["index"]
```
Comment (AM-USAF) [inline weave/pantry_factory.py:96]: Should you catch if there is no 'index' in the config file? Same with 'pantry_path' and 'file_system' or are we assuming the user will make their config correctly?

**Review by AM-USAF — CHANGES_REQUESTED** [review]
Great work! It all works _as expected_ but I think there is more we can do to catch errors for when people are using it not as expected. However, I'm new to all of this so if we don't prioritize that kind of testing on weave and you don't think catching those things is necessary for this ticket, then I am ready to approve.

**Comment by hh-usaf** [comment]
When I pass an index into create_pantry  with the pantry path but not the file system it still tries to do the third option. Should there be a different error code since the index isn't needed for the third, but the file system is? I may be a little lost

**Comment by hh-usaf** [comment]
I did not mean to do that

**Review by AM-USAF — CHANGES_REQUESTED** [review]
Do we need to update the readme with this new factory?

File: weave/pantry_factory.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+                os.path.join(kwargs.pop("pantry_path"), "config.json"),
+                'r'
+            ) as config_file:
+                config = json.load(config_file)
+        except FileNotFoundError as exc:
+            raise FileNotFoundError(
+                "Config file not found in pantry. Please add a config file "
+                "path or add a config file to the pantry."
+            ) from exc
+        pantry = _create_pantry_from_config(config, **kwargs)
```
Comment (hh-usaf) [inline weave/pantry_factory.py:86]: if the pantry does not exist this error is thrown which is not an accurate message for what is happening.

File: weave/tests/test_pantry_factory.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+        pantry_path=pantry_path,
+    )
+    index_name = test_index.index.__class__.__name__
+
+    yield test_pantry, index_name, index_constructor
+    test_pantry.cleanup_pantry()
+    test_index.cleanup_index()
+
+
+def test_pantry_factory_default_args(test_pantry):
```
Comment (hh-usaf) [inline weave/tests/test_pantry_factory.py:72]: would it be redundant to make tests with bad variables that are not in the config file?

File: README.md [hunk 6]
```diff
@@ -231,6 +231,48 @@ pantry = Pantry(
 )
 index_df = pantry.index.to_pandas_df()
 ```
+
+#### Pantry Factory
+Weave also has the ability to create a pantry from a config file using a pantry
+factory by calling `create_pantry()`. There are three ways of calling `create_pantry()`
+to enstantiate a pantry:
```
Comment (AM-USAF) [inline README.md:238]: instantiate

**Review by AM-USAF — APPROVED** [review]

**Review by hh-usaf — APPROVED** [review]

**Pull request merged**